contracts/zaps/ZapTrancheLp.sol

```diff
@@ -0,0 +1,277 @@
+        uint8 v;
+    }
+
+    /// @dev Takes the input permit calls and executes them
+    /// @param data The array which encodes the set of permit calls to make
+    modifier preApproval(PermitData[] memory data) {
+        // If permit calls are provided we make try to make them
+        if (data.length != 0) {
+            // We make permit calls for each indicated call
+            for (uint256 i = 0; i < data.length; i++) {
+                _permitCall(data[i]);
+            }
+        }
+        _;
+    }
+
+    /// @dev Makes permit calls indicated by a struct
+    /// @param data the struct which has the permit calldata
+    function _permitCall(PermitData memory data) internal {
+        // Make the permit call to the token in the data field using
+        // the fields provided.
+        // Security note - This fairly open call is safe because it cannot
+        // call 'transferFrom' or other sensitive methods despite the open
+        // scope. Do not make more general without security review.
+        data.tokenContract.permit(
+            msg.sender,
+            data.who,
+            data.amount,
+            data.expiration,
+            data.v,
+            data.r,
+            data.s
+        );
+    }
+
+    struct Out {
+        bytes32 poolId;
+        IVault.ExitPoolRequest request;
+    }
+
+    struct In {
+        uint256 lpCheck;
+        bytes32 poolId;
+        IVault.JoinPoolRequest request;
+    }
+
+    /// @notice Removes PT + YT liquidity from balancer pools, withdraws the PT + YT,
+    /// deposits into a new Tranche, and joins new PT and YT liquidity pools.
+    /// @dev The
+    /// @param _toMint The amount of underlying to mint with in the new tranche.
+    /// @param _ptOutInfo Out struct with info to exit the lp position for the pt.
+    /// @param _ytOutInfo Out struct with info to exit the lp position for the yt.
+    /// @param _ptInInfo In struct with info to enter the new lp position for the pt.
+    /// @param _ytInInfo In struct with info to enter the new lp position for the yt.
+    /// @param _onlyPrincipal This can be set to true to avoid any yt actions.
+    /// @param _permitCallData Encoded array of permit calls to make prior to minting
+    ///                        the data should be encoded with abi.encode(data, "PermitData[]")
+    ///                        each PermitData struct provided will be executed as a call.
+    ///                        An example use of this is if using a token with permit like USDC
+    ///                        to encode a permit which gives this contract allowance before minting.
+    function hopToTranche(
+        uint256 _toMint,
+        Out memory _ptOutInfo,
+        Out memory _ytOutInfo,
+        In memory _ptInInfo,
+        In memory _ytInInfo,
+        bool _onlyPrincipal,
+        PermitData[] calldata _permitCallData
+    ) public notFrozen {
+        IERC20 ytPool;
+        // get tranches
+        ITrancheExt trancheFrom = ITrancheExt(
```

Comment:
For auditablity we should add a link to the balancer code which is the the source for this logic